generators/chipyard/src/main/scala/IOBinders.scala

```diff
@@ -0,0 +1,167 @@
+package chipyard.iobinders
+
+import chisel3._
+
+import freechips.rocketchip.config.{Field, Config, Parameters}
+import freechips.rocketchip.diplomacy.{LazyModule}
+import freechips.rocketchip.devices.debug._
+import freechips.rocketchip.subsystem._
+import freechips.rocketchip.system._
+import freechips.rocketchip.util._
+
+import sifive.blocks.devices.gpio._
+import sifive.blocks.devices.uart._
+
+import testchipip._
+import icenet._
+import tracegen.{HasTraceGenTilesModuleImp}
+
+import scala.reflect.{ClassTag, classTag}
+
+// System for instantiating binders based
+// on the scala type of the Target (_not_ its IO). This avoids needing to
+// duplicate harnesses (essentially test harnesses) for each target.
+//
+// You could just as well create a custom harness module that instantiates
+// bridges explicitly, or add methods to
+// your target traits that instantiate the bridge there (i.e., akin to
+// SimAXI4Mem). Since cake traits live in Rocket Chip it was easiest to match
+// on the types rather than change trait code.
+
+
+
+// A map of partial functions that match on the type the DUT (_not_ it's
+// IO) to generate an appropriate bridge. You can add your own binder by adding
+// a new (key, fn) pair. You should override existing pairs in this map when
+// using a custom IOBinder
+
+// Since we also want to compose this structure like the existing config system,
+// use the scala string representation of the matched trait as a key
+
+case object IOBinders extends Field[Map[String, (Clock, Bool, Bool, Any) => Seq[Any]]](Map())
+
+
+// This macro overrides previous matches on some Top mixin. This is useful for
+// binders which modify IO, since those typically cannot be composed
+class RegisterIOBinder[T](fn: => (Clock, Bool, Bool, T) => Seq[Any])(implicit tag: ClassTag[T]) extends Config((site, here, up) => {
+  case IOBinders => up(IOBinders, site) + (tag.runtimeClass.toString ->
```

Comment:
Is this tag based on the class of the `IOBinder` itself, or the trait the user is matching on?

Reply:
The trait the user is matching on, which is determined from the function passed into this macro.